### DISCUSSION TITLE
SSD1362 256x64 driver Winstar WEA025664A

### DISCUSSION THREAD

**Discussion by provitek** [olikraus/u8g2#2043] [General]
Hi,
I wanted to make project with WInstar WEA025664A 256x64 oled display. SSD1362 controller.
unfortunately it is not directly compatible with u8g2.
Does anyone know what controller it is compatible with or closest to?
What would need to be changed?
Could it be added in the next release?
https://www.winstar.com.tw/pl/products/oled-module/graphic-oled-display/wea025664a.html
THX

**Comment by olikraus** [comment]
You could try the `U8G2_SSD1322_NHD_256X64_1_4W_SW_SPI` constructor. If you see something on the screen, please send a picture of the content. We could then use this constructor as a starting point for the creation of a SSD1362 constructor.

**Reply by provitek** [reply]
Thaank you.
I need couple of days to build my prototype. I wil try and send a picture.

**Reply by olikraus** [reply]
Remember to put u8g2 into 16 bit mode (see faq)

**Reply by provitek** [reply]
hi
I'm a novice programmer and I couldn't find U8G2_SSD1322_NHD_256X64_1_4W_SW_SPI. 
I found 3 initialization functions but I don't know what they are different. 
I used u8g2_Setup_ssd1322_nhd_256x64_2(&u8g2, U8G2_R0, u8x8_byte_stm32_hw_spi, u8x8_stm32_gpio_and_delay);
Forgive my ignorance.
I connected the display to NUCLEO F-401RE in hardware SPI mode.

The first tests of u8g2 in 16 bit mode and SSD1322 failed - the display was dark.
  After rewriting the test code that I got from the manufacturer, the display seems to work correctly (photo

 ![SSD1362_1](https://user-images.githubusercontent.com/50798040/207466045-4b46a200-753a-49d7-b20d-1f6b067a5c27.jpg)!
 (file OLEDtest.c, OLEDtest.h)!

So I modified the init sequence 8x8_d_ssd1322_256x64_init_seq[] to match what I got from Winstar and there was a little change - something appeared on the display but not what I wanted ;) photo
![SSD1362_2](https://user-images.githubusercontent.com/50798040/207466046-40d6fc35-40f5-4d50-b2be-c8041ca5f481.jpg)!

 instead of what is pictured
 ![SH1106](https://user-images.githubusercontent.com/50798040/207466041-2f172229-34a7-49a3-9a7d-6bac0edc2879.jpg)) !
[Disp_src.zip](https://github.com/olikraus/u8g2/files/10222850/Disp_src.zip)

Maybe I'm doing something wrong and it is not compatibility issue :)

THX
Chris

**Reply by olikraus** [reply]
Can you send a full u8g2 code of the none working example? Moreover: what changes did you exactly do? Can you send the modified values?

**Reply by provitek** [reply]
some files are in [Disp_src.zip](https://github.com/olikraus/u8g2/files/10222850/Disp_src.zip) (all modified + my src)
If you wish full project I will make git repository later.

**Comment by provitek** [comment]
full project as zip file 

I have modified  static const uint8_t u8x8_d_ssd1322_256x64_init_seq[]  in   u8x8_d_ssd1322.c and added u8g2_stm32f4 (.c + .h)
[OLEDtst-2.zip](https://github.com/olikraus/u8g2/files/10225193/OLEDtst-2.zip)

**Reply by olikraus** [reply]
I only have smartphone access only at the moment. So zip is difficult. Can you post the example code here. It shouldn't be more than 10 lines of code.

**Reply by olikraus** [reply]
I mainly need to see how you init and use u8g2.

**Reply by provitek** [reply]
I

``` C++
static const uint8_t u8x8_d_ssd1322_256x64_init_seq[] = {
    
  U8X8_DLY(1),
  U8X8_START_TRANSFER(),             	/* enable chip, delay is part of the transfer start */
  U8X8_DLY(1),
  
  
  U8X8_C(0xae),		                /* display off */
  U8X8_C(0XFD), //Set Command Lock
  U8X8_C(0X12), //(12H=Unlock,16H=Lock)

  U8X8_C(0XAE), //Display OFF(Sleep Mode)

  U8X8_C(0X15), //Set column Address
  U8X8_C(0X00), //Start column Address
  U8X8_C(0X7F), //End column Address

  U8X8_C(0X75), //Set Row Address
  U8X8_C(0X00), //Start Row Address
  U8X8_C(0X3F), //End Row Address

  U8X8_C(0X81), //Set contrast
  U8X8_C(0x2f),

  U8X8_C(0XA0), //Set Remap
  U8X8_C(0Xc3),

  U8X8_C(0XA1), //Set Display Start Line
  U8X8_C(0X00),

  U8X8_C(0XA2), //Set Display Offset
  U8X8_C(0X00),

  U8X8_C(0XA4), //Normal Display

  U8X8_C(0XA8), //Set Multiplex Ratio
  U8X8_C(0X3F),

  U8X8_C(0XAB), //Set VDD regulator
  U8X8_C(0X01), //Regulator Enable

  U8X8_C(0XAD), //External /Internal IREF Selection
  U8X8_C(0X8E),

  U8X8_C(0XB1), //Set Phase Length
  U8X8_C(0X22),

  U8X8_C(0XB3), //Display clock Divider
  U8X8_C(0XA0),

  U8X8_C(0XB6), //Set Second precharge Period
  U8X8_C(0X04),

  U8X8_C(0XB9), //Set Linear LUT

  U8X8_C(0XBc), //Set pre-charge voltage level
  U8X8_C(0X10), //0.5*Vcc

  U8X8_C(0XBD), //Pre-charge voltage capacitor Selection
  U8X8_C(0X01),

  U8X8_C(0XBE), //Set cOM deselect voltage level
  U8X8_C(0X07), //0.82*Vcc

  U8X8_C(0XAF), //Display ON

  U8X8_DLY(1),					/* delay 2ms */

  
  U8X8_END_TRANSFER(),             	/* disable chip */
  U8X8_END()             			/* end of sequence */
};
```

**Reply by provitek** [reply]
```c
        u8g2_Setup_ssd1322_nhd_256x64_2(&u8g2, U8G2_R0, u8x8_byte_stm32_hw_spi, u8x8_stm32_gpio_and_delay);
  	
/* Initialize Display */
  	u8g2_InitDisplay(&u8g2);
  	u8g2_SetPowerSave(&u8g2, 0);
  	White_IMAGE1(0xff);  //functions received from Winstar - all ON
  	HAL_Delay(3000);
  	White_IMAGE1(0x00);		//all off
  	HAL_Delay(3000);
  	White_IMAGE(*Image2);	//Text
  	HAL_Delay(3000);
		White_IMAGE1(0x00);
		HAL_Delay(3000);
  /* USER CODE END 2 */

  /* Infinite loop */
  /* USER CODE BEGIN WHILE */
  while (1)
  {
    /* USER CODE END WHILE */

    /* USER CODE BEGIN 3 */

	  u8g2_FirstPage(&u8g2);
	  do
	  {
		  u8g2_SetFont(&u8g2, u8g2_font_ncenB14_tr);
	      u8g2_DrawStr(&u8g2, 0, 15, "Hello World!");
	      u8g2_DrawCircle(&u8g2, 64, 40, 10, U8G2_DRAW_ALL);
	  } while (u8g2_NextPage(&u8g2));

  }

```

**Reply by olikraus** [reply]
Thanks for the code. So you say, that "USER CODE 2" works, but u8g2 "USER CODE 3" fails?

**Comment by provitek** [comment]
``` C++
#include "main.h"
#include "stm32f4xx_hal.h"
#include "u8g2.h"
#include "u8g2_stm32f4.h"

#define DEVICE_ADDRESS 	0x3C
#define TX_TIMEOUT		100

extern SPI_HandleTypeDef hspi2;
extern I2C_HandleTypeDef hi2c1;

uint8_t u8x8_stm32_gpio_and_delay(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)
{
	/* STM32 supports HW SPI, Remove unused cases like U8X8_MSG_DELAY_XXX & U8X8_MSG_GPIO_XXX */
	switch(msg)
	{
	case U8X8_MSG_GPIO_AND_DELAY_INIT:
		/* Insert codes for initialization */
		break;
	case U8X8_MSG_DELAY_MILLI:
		/* ms Delay */
		HAL_Delay(arg_int);
		break;
	case U8X8_MSG_GPIO_CS:
		/* Insert codes for SS pin control */
		HAL_GPIO_WritePin(OLED_CS_GPIO_Port, OLED_CS_Pin, arg_int);
		break;
	case U8X8_MSG_GPIO_DC:
		/* Insert codes for DC pin control */
		HAL_GPIO_WritePin(OLED_DC_GPIO_Port, OLED_DC_Pin, arg_int);
		break;
	case U8X8_MSG_GPIO_RESET:
		/* Insert codes for RST pin control */
		HAL_GPIO_WritePin(OLED_RST_GPIO_Port, OLED_RST_Pin, arg_int);
		break;
	}
	return 1;
}

uint8_t u8x8_byte_stm32_hw_spi(u8x8_t *u8x8, uint8_t msg, uint8_t arg_int, void *arg_ptr)
{
	switch(msg) {
	case U8X8_MSG_BYTE_SEND:
		/* Insert codes to transmit data */
		if(HAL_SPI_Transmit(&hspi2, arg_ptr, arg_int, TX_TIMEOUT) != HAL_OK) return 0;
		break;
	case U8X8_MSG_BYTE_INIT:
		/* Insert codes to begin SPI transmission */
		break;
	case U8X8_MSG_BYTE_SET_DC:
		/* Control DC pin, U8X8_MSG_GPIO_DC will be called */
		u8x8_gpio_SetDC(u8x8, arg_int);
		break;
	case U8X8_MSG_BYTE_START_TRANSFER:
		/* Select slave, U8X8_MSG_GPIO_CS will be called */
		u8x8_gpio_SetCS(u8x8, u8x8->display_info->chip_enable_level);
		HAL_Delay(1);
		break;
	case U8X8_MSG_BYTE_END_TRANSFER:
		HAL_Delay(1);
		/* Insert codes to end SPI transmission */
		u8x8_gpio_SetCS(u8x8, u8x8->display_info->chip_disable_level);
		break;
	default:
		return 0;
	}
	return 1;
}
```

**Reply by olikraus** [reply]
Is there any special need to use pure C here? I mean there is STM32duino for the Arduino IDE and U8g2 C++ Verision works without problems on STM32 with STM32duino. Especially you wouldn't need to write those HAL functions as they are part of U8g2 C++ Arduino version. But all in all your code looks ok.

**Reply by provitek** [reply]
I'n not familiar with STM32duino and debug the code and therefore use raw C and IDE from STM32. I will lern in the future STM32duino. THX :)

**Reply by olikraus** [reply]
Well I personally also prefer plain c code 😉

**Comment by olikraus** [comment]
Thanks for the code reply. I did a more closer look to both of the datasheets (SSD1322 vs SSD1362): Although the commands are almost the same, the SSD1362 doesn't need the 0x5c to enable RAM write:

https://github.com/olikraus/u8g2/blob/cd5b4704a873a375882917c97cd33f4470b379cb/csrc/u8x8_d_ssd1322.c#L205 

Moreover the logic level for the DC line for command arguments is inverted. As a result we need "cad_001" for SSD1362 (instead of "cad_011").

https://github.com/olikraus/u8g2/blob/cd5b4704a873a375882917c97cd33f4470b379cb/tools/codebuild/codebuild.c#L1847

I will create a complete new constructor for the SSD1362.

**Comment by olikraus** [comment]
I created an issue for this topic, let's continue there: https://github.com/olikraus/u8g2/issues/2051
